### DISCUSSION TITLE
Zig Zag

### DISCUSSION THREAD

**Discussion by DaveSkender** [DaveSkender/Stock.Indicators#226] [Indicators]
Community discussion about the [ZigZag](https://dotnet.stockindicators.dev/indicators/ZigZag/#content) indicator

![image](https://github.com/DaveSkender/Stock.Indicators/assets/8432125/2a5fb845-5565-42d6-88fc-915efb7c9ade)

**Comment by codebeaulieu** [comment]
I'm not sure how to interpret the output from this indicator. I'm not using any sort of drawing library to visualize the output, it looks like the output is useful for actually drawing. I write unit tests against the output results, but I cannot tell from the output, what the most recent zigZag high and low pivots are.

I'd expect to see something like your SuperTrend indicator where it's easy to tell based on the date and whether the upper band or lower band is null or not.

I understand algorithms differ across libraries, so I expect the output to be somewhat different. But if the inputs are similar, the values should be close. The way I work with indicators is by saving candle data to CSV files and then writing unit tests against them. So, for most indicators I can get very close to trading view charts.


<img width="1074" alt="tsla" src="https://user-images.githubusercontent.com/9411281/151698597-f83eb2c0-1486-4ffa-bd3a-36f07b03c71b.png">

I've tried three different input settings for the % change, 3.5, 5 and 10 and the output didn't appear to come remotely close to Trading View, or at least I couldn't interpret it correctly.

Here's the output for daily time interval, EndType.Close and percentChange of 5. The TradingView chart above had a deviation % of 5 and a depth of 10. Data source is polygon.io
```
---
retrace high: 1186.4142857142857142857142857
retrace low: 997.6476923076923076923076923
date: 01/04/2022 00:00:00
zigZag: 1156.575
---
retrace high: 1173.0485714285714285714285714
retrace low: 1007.4184615384615384615384615
date: 01/05/2022 00:00:00
zigZag: 1113.370
---
retrace high: 1159.6828571428571428571428571
retrace low: 1017.1892307692307692307692308
date: 01/06/2022 00:00:00
zigZag: 1070.165
---
retrace high: 1146.3171428571428571428571429
retrace low: 1026.96
date: 01/07/2022 00:00:00
zigZag: 1026.960
---
retrace high: 1132.9514285714285714285714286
retrace low: 1014.0592857142857142857142857
date: 01/10/2022 00:00:00
zigZag: 1053.38
---
retrace high: 1119.5857142857142857142857143
retrace low: 1001.1585714285714285714285714
date: 01/11/2022 00:00:00
zigZag: 1079.80
---
retrace high: 1106.2200000000000000000000000
retrace low: 988.2578571428571428571428571
date: 01/12/2022 00:00:00
zigZag: 1106.22
---
retrace high: 
retrace low: 975.3571428571428571428571429
date: 01/13/2022 00:00:00
zigZag: 1082.5954545454545454545454545
---
retrace high: 
retrace low: 962.4564285714285714285714286
date: 01/14/2022 00:00:00
zigZag: 1058.9709090909090909090909091
---
retrace high: 
retrace low: 949.5557142857142857142857143
date: 01/18/2022 00:00:00
zigZag: 1035.3463636363636363636363636
---
retrace high: 
retrace low: 936.6550000000000000000000000
date: 01/19/2022 00:00:00
zigZag: 1011.7218181818181818181818182
---
retrace high: 
retrace low: 923.7542857142857142857142857
date: 01/20/2022 00:00:00
zigZag: 988.0972727272727272727272727
---
retrace high: 
retrace low: 910.8535714285714285714285714
date: 01/21/2022 00:00:00
zigZag: 964.4727272727272727272727273
---
retrace high: 
retrace low: 897.9528571428571428571428571
date: 01/24/2022 00:00:00
zigZag: 940.8481818181818181818181818
---
retrace high: 
retrace low: 885.0521428571428571428571428
date: 01/25/2022 00:00:00
zigZag: 917.2236363636363636363636364
---
retrace high: 
retrace low: 872.1514285714285714285714286
date: 01/26/2022 00:00:00
zigZag: 893.5990909090909090909090909
---
retrace high: 
retrace low: 859.2507142857142857142857143
date: 01/27/2022 00:00:00
zigZag: 869.9745454545454545454545454
---
retrace high: 
retrace low: 846.3500000000000000000000000
date: 01/28/2022 00:00:00
zigZag: 846.3500000000000000000000000```

### Questions

Could you please help me to understand the output. 
Where are the pivot points from this indicator? If the zigZag value represents the pivot points, there seems to be too many, even if I have the percent change set to 10 there are pivots nearly every day.

Let me know if you need clarifications or any more info from me.

**Reply by DaveSkender** [reply]
The first thing I'm seeing here is that the chart is pivoting on High/Low, not Close, so to be the same, you'd have to use `EndType.HighLow` as your parameter.

The first span is always unreliable too since it has to find the first pivot, so you should ignore that section.  There are some things to also consider if the underlying data is different too.

**Reply by DaveSkender** [reply]
Another thing to note, the "zigzag" value in your results is the actual line between points; so to isolate the actual Pivots, you'd have to filter it to only include `PivotType` is `H` or `L`, or simply not `null`.

**Comment by codebeaulieu** [comment]
@DaveSkender Relevant issue for the Stock.Charts library. https://github.com/DaveSkender/Stock.Charts/issues/180

**Reply by DaveSkender** [reply]
Yep, I'm actively working towards getting the demo site updated to shows more current chart data for a few symbols and for all indicators in the library, to help people visualize what these actually look like in a live-action scenario.

I suppose it could also be used to compare with other charting sites' output; though, you have to also consider all the factors to understand why they might be different -- including that the other site might be wrong!  😃

**Reply by DaveSkender** [reply]
You can see https://github.com/DaveSkender/Stock.Charts/pull/181 as an example of how I added ZigZag to our demo chart.

**Comment by M22arius494** [comment]
Hello Dave,

I'm also observing that the results of the ZigZag are rather different than what TradingView is presenting/calculating. Yours is picking up additional highs and lows which TradingView is ignoring. TradingView is using a depth which isn't in you r sources (not sure if it should be).

For example the AXSBTC@Binance. I'm using the 2h candles from 1 december 21 up to 3 march 22 (746 candles should be more than enough) and start comparing from 1 jan 22. The first 2 are perfect but the 3 and 3th are weird (the two marked with the question marks). Because of this extra h/l the ZigZag is unusable for my needs (the results from this example isn't to bad though). In other cases it will pick up additional lower lows and lower highs while in fact it is properly trending on TV.


date=01-01-2022 11:00:00 zigzag=0.00194800  PointType=L  RetraceHigh=  RetraceLow=0.00194800
date=03-01-2022 15:00:00 zigzag=0.00207400  PointType=H  RetraceHigh=0.00207400  RetraceLow=0.00186230

date=05-01-2022 23:00:00 zigzag=0.00177000  PointType=L  RetraceHigh=0.00191853  RetraceLow=0.00177000  ???
date=06-01-2022 19:00:00 zigzag=0.00186300  PointType=H  RetraceHigh=0.00186300  RetraceLow=0.00172467 ???

date=07-01-2022 05:00:00 zigzag=0.00170200  PointType=L  RetraceHigh=0.00184682  RetraceLow=0.00170200
date=08-01-2022 05:00:00 zigzag=0.00180800  PointType=H  RetraceHigh=0.00180800  RetraceLow=0.00168165
date=11-01-2022 01:00:00 zigzag=0.00162400  PointType=L  RetraceHigh=0.00180615  RetraceLow=0.00162400
date=12-01-2022 19:00:00 zigzag=0.00180500  PointType=H  RetraceHigh=0.00180500  RetraceLow=0.00165741

@code
decimal zigZagPercentage = 5;
List<ZigZagResult> resultZigZag = (List<ZigZagResult>)Indicator.GetZigZag(candles.Values.ToList(), EndType.HighLow, zigZagPercentage);
//^The editor is removing some generics in above line

for (int i = 0; i < resultZigZag.Count; i++)
{
  ZigZagResult zigZagResult = resultZigZag[i];

  string s = string.Format("date={0} zigzag={1:N8}  PointType={2:N8}  RetraceHigh={3:N8}  RetraceLow={4:N8}", zigZagResult.Date.ToLocalTime(), zigZagResult.ZigZag, zigZagResult.PointType, zigZagResult.RetraceHigh, zigZagResult.RetraceLow);
  AddTextToLogTab(s);
}

![image](https://user-images.githubusercontent.com/59998287/160253678-6aa94773-1f91-4ddc-a960-571e8c6f1a89.png)

**Reply by DaveSkender** [reply]
Thanks for your feedback, Marius.  I think adding this new `depth` parameter can help users skip short-term reversals.  This is already in our backlog.  See #694.  Currently, if you use a `Depth` value of `1` in TradingView, it should look the same, assuming their base approach is the same.

Our current formula is the same one used by [StockCharts.com](https://school.stockcharts.com/doku.php?id=technical_indicators:zigzag), Fidelity, and most other platforms.

Just as an FYI, as a rule, we don't source our formulas from TradingView or any other third-party source code because it's impossible to validate them against reputable original sources of formulas.  In other words, we can’t know whether they are implemented correctly.  Sometimes they are not.  This is especially true for community-contributed indicators as there is little or no quality control on them.

**Comment by Gadirino** [comment]
Hello @DaveSkender 
I used ((( indicators.get_zig_zag )))  in my python codes and the outputs were very irregular. 
What do you think is the problem?
There are tow different chart data and zigzag outputs:
![z1](https://user-images.githubusercontent.com/110600525/182823270-decca4f7-e937-42c9-bb27-3a65a1b2252c.jpg)
![z2](https://user-images.githubusercontent.com/110600525/182823279-770a7306-6874-449b-ab60-048b94fe5d06.jpg)

**Reply by DaveSkender** [reply]
It's hard to say.  It almost looks like you're overlaying results from a different ticker symbol.  @LeeDongGeon1996 any thoughts?

**Reply by LeeDongGeon1996** [reply]
~Are those calculated from same data? and got different results?~ I get it.
Umm.. because of passing too small `percent_change` value, that is unable to ignore small percent movement?

**Reply by Gadirino** [reply]
Thanks a lot for your answers
I resolved that. 
My input data have some unknown problems in its date column. I changed data to another quote and the problem solved.

**Comment by Abdulhadi77** [comment]
How to draw the zigzag in python?

**Reply by DaveSkender** [reply]
I don't have any specific ideas for Python, but checkout our list of [charting and visualization options](https://github.com/DaveSkender/Stock.Indicators/discussions/430)

**Reply by LeeDongGeon1996** [reply]
I recommend [Plotly for Python](https://plotly.com/python/).

**Comment by adealkdke** [comment]
How do I get my zigzag indicator with pivots to look like yours from your website stock indicators.dev specifically referencing the zigzag indicator. I have tried multiple times to replicate your chart indicator but have not been able to get straight lines like yours from the pivot reversals and the zigzag lines.

**Reply by DaveSkender** [reply]
Start with a standard 5% `percentChange` deviation with high/low `endType` reversals.  If your prices have typical volatility it should look something like this.

```csharp
// C# usage 
var results = quotes.GetZigZag(
  endType: EndType.HighLow,
  percentChange: 5);
```

![image](https://github.com/user-attachments/assets/9625a92e-bf41-4138-8323-0266f2647729)

If you’re seeing too many or too few reversals, increase or decrease the change threshold. You can [play around with different values](https://charts.stockindicators.dev/) to see what they do.

If you’re plotting on your own chart, and are seeing flat, jagged, or missing spots in the lines between pivots, ensure your axis scale isn’t inserting extra days that don’t correspond to your `quotes` — some time scales will do that.  Most market prices won’t include weekends.

**Reply by DaveSkender** [reply]
Also, check the dates on your `quotes`.  Like the [example above](https://github.com/DaveSkender/Stock.Indicators/discussions/226#discussioncomment-3324044), bad or incorrectly sorted dates can cause problems.  We have a `quotes.Validate()` utility that can check those.

**Comment by ydqsbfsyi** [comment]
I'm getting this error: No method matches given arguments for List`1.Add: (<class 'str'>)
I checked, "date" dtype is datetime64[ns], like yy-mm-dd hh:mm:ss, all others are float64.
I tried "quotes.Validate()", but I got AttributeError: 'DataFrame' object has no attribute 'Validate'.
Could anybody please give me a hint?

**Reply by DaveSkender** [reply]
This doesn't sound like an issue with ZigZag, but rather a more basic issue with how you're composing your quotes from another Dataframe object.  There are examples on composing quotes in our [Quickstart guide](https://github.com/facioquo/stock-indicators-python-quickstart) that may help.

**Reply by ydqsbfsyi** [reply]
> This doesn't sound like an issue with ZigZag, but rather a more basic issue with how you're composing your quotes from another Dataframe object. There are examples on composing quotes in our [Quickstart guide](https://github.com/facioquo/stock-indicators-python-quickstart) that may help.

Thank you very much for your help.  I'll look into it.

**Comment by Paroca72** [comment]
Hello,

I'm using the C# library and filtered the result on the last 100 period.
I have as result many points whole null.. that is not a big problem because I can filter on it.
Mostly important the only one field, sometime initialized, is "ZigZag".. all other fields are always null.

<img width="1742" height="942" alt="image" src="https://github.com/user-attachments/assets/cef9c7c4-934d-4c9c-b59b-240a3e160283" />

UPDATE:
I also try to change the percentage and also not filter on the last 100 period but use the whole candle list (more than 1400) but the PointType is always null

<img width="1678" height="919" alt="image" src="https://github.com/user-attachments/assets/ffcb39fc-4c6a-4bd3-8710-70cf54152c3e" />

Thanks for your help
Sam

**Reply by DaveSkender** [reply]
There are some situations where the first periods or all periods will be `null` depending on the quotes and params you provide

<img width="1488" height="695" alt="image" src="https://github.com/user-attachments/assets/712be692-188e-486f-af2c-bcb23258b15c" />

---

example: pivot never detected (`percentChange` too high)

<img width="1488" height="697" alt="image" src="https://github.com/user-attachments/assets/f4af3d21-ab23-47dc-adcf-e07cb6062a43" />

**Reply by DaveSkender** [reply]
Upload the raw quotes and param values you’re using for `quotes.GetZigZag(endType, percentChange)` to a [new issue](https://github.com/DaveSkender/Stock.Indicators/issues/new) and I’ll try to repro.  Just guessing based on what your showing, there’s likely an issue in your own code, which I can’t see fully to troubleshoot, but can try to rule out actual library issues.

**Reply by DaveSkender** [reply]
See the following issue with a recommendation on how to fix this in your implementation:

- #1949 (user support)
